qiskit/circuit/parameterexpression.py

```
@@ -613,28 +612,19 @@ def __eq__(self, other):
         if isinstance(other, ParameterExpression):
             if self.parameters != other.parameters:
                 return False
-            from sympy import sympify
-
-            if not isinstance(self._symbol_expr, symengine.Basic):
-                raise QiskitError("Invalid ParameterExpression")
 
-            return sympify(self._symbol_expr).equals(sympify(other._symbol_expr))
+            return self._symbol_expr == other._symbol_expr
         elif isinstance(other, numbers.Number):
-            return len(self.parameters) == 0 and complex(self._symbol_expr) == other
+            return self._symbol_expr == other
         return False
 
     def is_real(self):
         """Return whether the expression is real"""
-        if not self._symbol_expr.is_real and self._symbol_expr.is_real is not None:
-            # Symengine returns false for is_real on the expression if
-            # there is a imaginary component (even if that component is 0),
-            # but the parameter will evaluate as real. Check that if the
-            # expression's is_real attribute returns false that we have a
-            # non-zero imaginary
-            if self._symbol_expr.imag == 0.0:
-                return True
-            return False
-        return self._symbol_expr.is_real
+        try:
+            val = self._symbol_expr.value()
+            return not isinstance(val, complex)
+        except RuntimeError:
```

I think it's better if we can catch a more-specific error type here. If something bad happens that we don't expect, then it's better to let the exception bubble all the way up.

I see that when working with symengine we needed to catch `RuntimeError`, but that should no longer be the case.